### PR TITLE
Add React Source and Layer components (PR 1/2)

### PR DESCRIPTION
**Pull request by Pessimistress**
See documentation

### CODE REVIEW

File: src/components/source.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        sourceOptions[key] = props[key];
+        changedKey = key;
+        changedKeyCount++;
+      }
+    }
+
+    const {type, _map: map} = this;
+    const source = this.getSource();
+    if (!source) {
+      this._createSource();
```
Comment (xintongxia) [inline src/components/source.js]: Move this creating source before checking source option changes.

File: src/components/layer.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import type {MapContextProps} from './map-context';
+
+const propTypes = {
+  type: PropTypes.string.isRequired,
+  id: PropTypes.string,
+  source: PropTypes.string,
+  beforeId: PropTypes.string
+};
+
+type LayerProps = {
```
Comment (xintongxia) [inline src/components/layer.js]: Mapox layer has other props, so will this type spec cause any flow errors if we are gonna publish flow definition?

File: src/components/source.js (hunk 1 above)
Comment (Pessimistress) [reply src/components/source.js]: `_sourceOptions` is populated in the for loop.

File: src/components/layer.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    if (map.style._loaded) {
+      // console.log('adding layer');
+      map.addLayer(options, this.props.beforeId);
+    } else {
+      map.once('styledata', () => this.forceUpdate());
+    }
+  }
+
+  /* eslint-disable complexity */
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply src/components/layer.js]: Yes.

File: src/components/layer.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    try {
+      diffLayerStyles(map, this.id, props, prevProps);
+    } catch (error) {
+      console.warn(error); // eslint-disable-line
+    }
+  }
+  /* eslint-disable complexity */
+
+  _render(context: MapContextProps) {
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply src/components/layer.js]: To save the context?

File: src/components/layer.js (hunk 2 above)
Comment (Pessimistress) [reply src/components/layer.js]: I don't think so: https://flow.org/try/#0C4TwDgpgBAglC8UDeAoKUCGAuKA7ArgLYBGEATigL4DcKKAbhmVAPbEBWAjFnIktlADkgygyasOAJh4JkAgAwAaKMRzDlAYxzAy+CKKA

File: src/components/source.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+      source.updateImage({url: sourceOptions.url, coordinates: sourceOptions.coordinates});
+    } else if (
+      (type === 'canvas' || type === 'video') &&
+      changedKeyCount === 1 &&
+      changedKey === 'coordinates'
+    ) {
+      source.setCoordinates(sourceOptions.coordinates);
+    } else {
+      map.removeSource(this.id);
+      map.addSource(sourceOptions);
```
Comment (VladimirShumchenia) [inline src/components/source.js]: @Pessimistress I think map.addSource(sourceOptions); => map.addSource(**this.id**, this._sourceOptions)

Comment (Pessimistress) [reply src/components/source.js]: Thanks. Fixed in https://github.com/uber/react-map-gl/pull/923

**Pull request merged**